actors/miner/src/lib.rs

```diff
@@ -4962,6 +4946,139 @@ fn confirm_sector_proofs_valid_internal(
+                    "error activating deals on sector {}: {}",
+                    pre_commit.info.sector_number,
+                    e.msg()
+                );
+                if all_or_nothing {
+                    return Ok(None);
+                } else {
+                    continue;
+                }
+            }
+        };
+
+        activation_results.push((pre_commit, activate_res));
+    }
+
+    // When all prove commits have failed abort early
+    if activation_results.is_empty() {
+        return Err(actor_error!(illegal_argument, "all sectors failed to activate"));
+    }
+
+    // Claims aggregated across all sectors
+    let mut sectors_claims = Vec::new();
+    for (pre_commit, activate_res) in activation_results.clone() {
+        let mut sector_claims = activate_res
+            .verified_infos
+            .iter()
+            .map(|info| ext::verifreg::SectorAllocationClaim {
+                client: info.client,
+                allocation_id: info.allocation_id,
+                data: info.data,
+                size: info.size,
+                sector: pre_commit.info.sector_number,
+                sector_expiry: pre_commit.info.expiration,
+            })
+            .collect();
+        sectors_claims.append(&mut sector_claims);
+    }
+
+    if sectors_claims.is_empty() {
+        Ok(Some(
+            activation_results
+                .iter()
+                .map(|(pre_commit, activation_result)| {
+                    (
+                        pre_commit.clone(),
+                        ext::market::DealSpaces {
+                            verified_deal_space: BigInt::default(),
+                            deal_space: activation_result.nonverified_deal_space.clone(),
+                        },
+                    )
+                })
+                .collect(),
+        ))
+    } else {
+        let claim_raw = extract_send_result(rt.send_simple(
+            &VERIFIED_REGISTRY_ACTOR_ADDR,
+            ext::verifreg::CLAIM_ALLOCATIONS_METHOD,
+            IpldBlock::serialize_cbor(&ext::verifreg::ClaimAllocationsParams {
+                sectors: sectors_claims,
+                all_or_nothing: true,
+            })?,
+            TokenAmount::zero(),
+        ));
+
+        let claim_res: ext::verifreg::ClaimAllocationsReturn = match claim_raw {
+            Ok(res) => deserialize_block(res)?,
+            Err(e) => {
+                info!("error claiming allocations for batch {}", e.msg());
+                return Ok(None);
+            }
+        };
+
+        let activation_and_claim_results: Vec<(
+            SectorPreCommitOnChainInfo,
+            ext::market::DealSpaces,
+        )> = activation_results
+            .iter()
+            .map(|(pre_commit, activation_result)| {
+                let verified_deal_space = claim_res
+                    .claim_results
+                    .iter()
+                    .filter(|c| pre_commit.info.sector_number == c.sector)
```

Comment:
After making the claim results parallel to sectors_claims, we can group by sector with a single iteration rather than this quadratic.